src/app/auth/permission.service.ts

```diff
@@ -8,54 +12,134 @@ export type Permission = {
   scopes: Array<string>
 };
 
+export type UserRoleData = {
+  assignee_id: string,
+  assignee_type: string,
+  inherited: boolean,
+  inherited_from: string,
+  role_name: string
+};
+
 @Injectable()
 export class PermissionService {
   private jwtHelper: JwtHelperService = new JwtHelperService();
+  private headers = new HttpHeaders({ 'Content-Type': 'application/json' });
+
+  constructor(
+    private http: HttpClient,
+    @Inject(AUTH_API_URL) private authApi: string
+  ) {}
 
   /**
    * Returns all the scopes a user has for a specific resource.
    * @param resourceId ID of a specific resource such as a Space
    */
-  getAllScopes(resourceId: string): Array<string> {
-    const permissions = this.getPermission(resourceId);
-    return permissions ? permissions.scopes : [];
+  getAllScopes(resourceId: string): Observable<string[]> {
+    return this.getPermission(resourceId).pipe(
+      map((permission: Permission) => permission.scopes)
+    );
   }
 
   /**
    * Checks if a user has a specific scope for a resource.
    * @param resourceId ID of a specific resource such as a Space
    * @param scope the scope you want to check for. Ex - `can edit`
    */
-  hasScope(resourceId: string, scope: string): boolean {
-    const permissions = this.getPermission(resourceId);
-    return permissions ? permissions.scopes.includes(scope) : false;
+  hasScope(resourceId: string, scope: string): Observable<boolean> {
+    return this.getPermission(resourceId).pipe(
+      map((permission: Permission) => permission.scopes.includes(scope))
+    );
   }
 
   /**
    * Returns the permission for a specific resource.
+   * If there is no permission info in current RPT then we need to audit new RPT token.
    * @param resourceId ID of a specific resource such as a Space
    */
-  getPermission(resourceId: string): Permission | null {
-    const decodedToken = this.getDecodedToken();
-    if (this.isValidRPT(decodedToken)) {
-      return decodedToken.permissions.find((permission: Permission) => permission.resource_set_id === resourceId);
-    }
-    return null;
+  getPermission(resourceId: string): Observable<Permission | undefined> {
+    const permission = this.findPermission(this.getDecodedToken(), resourceId);
+    return iif(
+      () => !!permission,
+      of(permission),
+      this.findPermissionAfterAudit(resourceId)
+    );
   }
 
   /**
-   * Decodes the JWT token using JwtHelperService from `angular-jwt`.
+   * Assign a specific role to a one or more users in a resource such as space.
+   * @param resourceId ID of a specific resource such as a Space
+   * @param roleName Role to be assigned for the resource
+   * @param userIDs IDs for users that need to be assigned the specified role.
    */
-  private getDecodedToken(): any {
-    const token = localStorage.getItem('auth_token');
-    return token ? this.jwtHelper.decodeToken(token) : '';
+  assignRole(resourceId: string, roleName: string, userIDs: Array<string>): Observable<any> {
```

Comment:
Why does this return `Observable<any>` and not an observable of a specific value or `undefined` | `null` ?